pkg/approver/aws/csr.go

```diff
@@ -0,0 +1,184 @@
+	if err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("handle: running recognizers on %s", csr.GetName())
+	csrrs := ar.recognizers()
+	for _, csrr := range csrrs {
+		rs := csrr.recognizers
+		approved := true
+		for _, r := range rs {
+			if !r(csr, x509cr) {
+				approved = false
+				break
+			}
+		}
+		if !approved {
+			continue
+		}
+
+		glog.V(4).Infof("csr %s was approved! message: %s", csr.GetName(), csrr.successMessage)
+		csr.Status.Conditions = append(csr.Status.Conditions, certificates.CertificateSigningRequestCondition{
+			Type:    certificates.CertificateApproved,
+			Reason:  "AutoApproved",
+			Message: csrr.successMessage,
+		})
+		_, err = ar.kubeClient.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(csr)
+		if err != nil {
+			return fmt.Errorf("error updating approval for csr: %v", err)
+		}
+		break
+	}
+
+	return nil
+}
+
+type recognizerFunc func(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool
+type csrRecognizer struct {
+	recognizers    []recognizerFunc
+	successMessage string
+}
+
+func (ar *Approver) recognizers() []csrRecognizer {
+	return []csrRecognizer{{
+		// ensures: node present their client cert, exist(running) in cluster, belong to whitelist of ASGs.
+		recognizers:    []recognizerFunc{isSelfNodeClientCert, ar.isValidNode, ar.isValidASG},
+		successMessage: "kube-aws-approver approved self node client cert",
+	}, {
+		// ensures: node username has instance-id, presented instance-id match instance-id by aws for node name in running state, doesn't belong to cluster, belong to whitelist of ASGs.
+		recognizers:    []recognizerFunc{isNodeClientCert, ar.isValidNewNode, ar.isValidASG},
+		successMessage: "kube-aws-approver approved new node client cert",
+	}}
+}
+
+func isNodeClientCert(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !reflect.DeepEqual([]string{"system:nodes"}, x509cr.Subject.Organization) {
+		return false
+	}
+	if (len(x509cr.DNSNames) > 0) || (len(x509cr.EmailAddresses) > 0) || (len(x509cr.IPAddresses) > 0) {
+		return false
+	}
+	if !hasExactUsages(csr, kubeletClientUsages) {
+		return false
+	}
+	if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
+		return false
+	}
+	return true
+}
+
+func isSelfNodeClientCert(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeClientCert(csr, x509cr) {
+		return false
+	}
+	if csr.Spec.Username != x509cr.Subject.CommonName {
+		return false
+	}
+	return true
+}
+
+func (ar *Approver) isValidNewNode(csr *certificates.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	idu := getInstanceIDFromUsername(csr.Spec.Username)
+	nn := getNodeNameFromCN(x509cr.Subject.CommonName)
+	idn, err := ar.aws.instanceID(nn)
+	if err != nil {
+		return false
+	}
+	if idn != idu {
+		return false
+	}
+	_, err = ar.kubeClient.CoreV1().Nodes().Get(nn, metav1.GetOptions{})
```

Comment:
What is the node was previously deleted, because it was compromised? Need to at least document.

Reply:
Would you recommend i add a check that there is no already approved CertificateSigningRequest for that nodename.
Or add this to comment
```
// If a malicious code deletes the node object from kubernetes API and requests for a new client cert from the same node, it will get approved for new node.
```